File: src/main/java/uk/co/eelpieconsulting/feedlistener/controllers/ChannelsController.java
package uk.co.eelpieconsulting.feedlistener.controllers;

import com.google.common.base.Strings;
import com.mongodb.MongoException;
import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;
import uk.co.eelpieconsulting.common.views.ViewFactory;
import uk.co.eelpieconsulting.feedlistener.IdBuilder;
import uk.co.eelpieconsulting.feedlistener.UrlBuilder;
import uk.co.eelpieconsulting.feedlistener.daos.ChannelsDAO;
import uk.co.eelpieconsulting.feedlistener.daos.FeedItemDAO;
import uk.co.eelpieconsulting.feedlistener.daos.SubscriptionsDAO;
import uk.co.eelpieconsulting.feedlistener.daos.UsersDAO;
import uk.co.eelpieconsulting.feedlistener.exceptions.UnknownUserException;
import uk.co.eelpieconsulting.feedlistener.model.Channel;
import uk.co.eelpieconsulting.feedlistener.model.FeedItemsResult;
import uk.co.eelpieconsulting.feedlistener.model.Subscription;

import javax.servlet.http.HttpServletResponse;
import java.net.UnknownHostException;
import java.util.List;

@Controller
public class ChannelsController {

    private static final Logger log = Logger.getLogger(ChannelsController.class);

    private static final String X_TOTAL_COUNT = "X-Total-Count";

    private final UsersDAO usersDAO;
    private final ChannelsDAO channelsDAO;
    private final SubscriptionsDAO subscriptionsDAO;
    private final IdBuilder idBuilder;
    private final UrlBuilder urlBuilder;
    private final ViewFactory viewFactory;
    private final FeedItemPopulator feedItemPopulator;
    private final FeedItemDAO feedItemDAO;

    @Autowired
    public ChannelsController(UsersDAO usersDAO, ChannelsDAO channelsDAO, SubscriptionsDAO subscriptionsDAO,
                              IdBuilder idBuilder, UrlBuilder urlBuilder, ViewFactory viewFactory,
                              FeedItemPopulator feedItemPopulator, FeedItemDAO feedItemDAO) {
        this.usersDAO = usersDAO;
        this.channelsDAO = channelsDAO;
        this.subscriptionsDAO = subscriptionsDAO;
        this.idBuilder = idBuilder;
        this.urlBuilder = urlBuilder;
        this.viewFactory = viewFactory;
        this.feedItemPopulator = feedItemPopulator;
        this.feedItemDAO = feedItemDAO;
    }

    @RequestMapping(value = "/{username}/channels", method = RequestMethod.GET)
    public ModelAndView channelsJson(@PathVariable String username) throws UnknownUserException {
        usersDAO.getByUsername(username);

        log.info("Channels for user: " + username);
        return new ModelAndView(viewFactory.getJsonView()).
                addObject("data", channelsDAO.getChannels(username));
    }

    @RequestMapping(value = "/{username}/channels/{id}", method = RequestMethod.GET)
    public ModelAndView channel(@PathVariable String username, @PathVariable String id) throws UnknownHostException, MongoException, UnknownUserException {
        usersDAO.getByUsername(username);

        final Channel channel = channelsDAO.getById(username, id);

        return new ModelAndView(viewFactory.getJsonView()).
                addObject("data", channel);
    }

    @RequestMapping(value = "/{username}/channels/{id}/subscriptions", method = RequestMethod.GET)
    public ModelAndView channelSubscriptions(@PathVariable String username, @PathVariable String id, @RequestParam(required = false) String url) throws MongoException, UnknownUserException {
        usersDAO.getByUsername(username);

        final Channel channel = channelsDAO.getById(username, id);
        final List<Subscription> subscriptionsForChannel = subscriptionsDAO.getSubscriptionsForChannel(username, channel.getId(), url);

        final ModelAndView mv = new ModelAndView(viewFactory.getJsonView());
        mv.addObject("data", subscriptionsForChannel);
        return mv;
    }

    @RequestMapping(value = "/{username}/channels/{id}/items", method = RequestMethod.GET)
    public ModelAndView channelJson(@PathVariable String username, @PathVariable String id,
                                    @RequestParam(required = false) Integer page,
                                    @RequestParam(required = false) Integer pageSize,
                                    @RequestParam(required = false) String format,
                                    @RequestParam(required = false) String q,
                                    HttpServletResponse response) throws MongoException, UnknownUserException {
        usersDAO.getByUsername(username);
        final Channel channel = channelsDAO.getById(username, id);

        ModelAndView mv = new ModelAndView(viewFactory.getJsonView());
        if (!Strings.isNullOrEmpty(format) && format.equals("rss")) {    // TODO view factory could do this?
            mv = new ModelAndView(viewFactory.getRssView(channel.getName(), urlBuilder.getChannelUrl(channel), ""));
        }

        FeedItemsResult results = feedItemDAO.getChannelFeedItemsResult(username, channel, page, q, pageSize);
        feedItemPopulator.populateFeedItems(results, mv, "data");
        long totalCount = results.getTotalCount();
        response.addHeader(X_TOTAL_COUNT, Long.toString(totalCount));
        return mv;
    }

    @RequestMapping(value = "/{username}/channels", method = RequestMethod.POST)
    public ModelAndView addChannel(@PathVariable String username, @RequestParam String name) throws UnknownUserException {
        usersDAO.getByUsername(username);

        Channel newChannel = new Channel(idBuilder.makeIdFor(name), name, username);
        channelsDAO.add(username, newChannel);

        return new ModelAndView(new RedirectView(urlBuilder.getChannelUrl(newChannel)));
    }

}

File: src/main/kotlin/uk/co/eelpieconsulting/feedlistener/controllers/ChannelsController.kt
package uk.co.eelpieconsulting.feedlistener.controllers

import com.google.common.base.Strings
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.apache.logging.log4j.LogManager
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.servlet.ModelAndView
import uk.co.eelpieconsulting.common.views.ViewFactory
import uk.co.eelpieconsulting.feedlistener.IdBuilder
import uk.co.eelpieconsulting.feedlistener.UrlBuilder
import uk.co.eelpieconsulting.feedlistener.controllers.ui.WithSignedInUser
import uk.co.eelpieconsulting.feedlistener.daos.ChannelsDAO
import uk.co.eelpieconsulting.feedlistener.daos.FeedItemDAO
import uk.co.eelpieconsulting.feedlistener.daos.SubscriptionsDAO
import uk.co.eelpieconsulting.feedlistener.model.Channel
import uk.co.eelpieconsulting.feedlistener.model.User

@Controller
class ChannelsController @Autowired constructor(private val channelsDAO: ChannelsDAO,
                                                private val viewFactory: ViewFactory,
                                                val subscriptionsDAO: SubscriptionsDAO,
                                                val urlBuilder: UrlBuilder,
                                                private val feedItemPopulator: FeedItemPopulator,
                                                private val feedItemDAO: FeedItemDAO,
                                                private val conditionalLoads: ConditionalLoads,
                                                private val idBuilder: IdBuilder,
                                                currentUserService: CurrentUserService,
                                                request: HttpServletRequest) : WithSignedInUser(currentUserService, request) {

    private val log = LogManager.getLogger(ChannelsController::class.java)

    private val X_TOTAL_COUNT = "X-Total-Count"

    @CrossOrigin
    @GetMapping("/{username}/channels")
    fun channelsJson(@PathVariable username: String): ModelAndView {
        fun renderChannels(user: User): ModelAndView {
            if (user.username != username) {
                throw ResponseStatusException(HttpStatus.FORBIDDEN, "Cannot list another users channels")
            }
            return ModelAndView(viewFactory.getJsonView()).addObject("data", channelsDAO.getChannelsFor(user))
        }
        return forCurrentUser(::renderChannels)
    }

    @CrossOrigin
    @GetMapping("/channels/{id}")
    fun channel(@PathVariable id: String): ModelAndView {
        return forCurrentUser { user ->
            conditionalLoads.withChannelForUser(id, user) { channel ->
                ModelAndView(viewFactory.getJsonView()).addObject("data", channel)
            }
        }
    }

    @CrossOrigin
    @PostMapping("/channels")
    fun createSubscription(@RequestBody create: CreateChannelRequest): ModelAndView {
        fun createChannel(user: User): ModelAndView {
            log.info("Got channel create request: " + create)
            if (Strings.isNullOrEmpty(create.name)) {
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Channel name is required")
            }

            if (channelsDAO.usersChannelByName(user, create.name) == null) {
                val newChannel = Channel(id = idBuilder.makeIdForChannel(), name = create.name, username = user.username)
                channelsDAO.save(newChannel)
                log.info("Added channel: $newChannel")
                return ModelAndView(viewFactory.getJsonView()).addObject("data", newChannel)
            } else {
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Cannot create channel for anonymous user")
            }
        }

        return forCurrentUser(::createChannel)
    }

    @CrossOrigin
    @GetMapping("/channels/{id}/subscriptions")
    fun channelSubscriptions(@PathVariable id: String, @RequestParam(required = false) url: String?): ModelAndView {
        return forCurrentUser { user ->
            conditionalLoads.withChannelForUser(id, user) { channel ->
                val subscriptionsForChannel = subscriptionsDAO.getSubscriptionsForChannel(channel.id, url)
                ModelAndView(viewFactory.getJsonView()).addObject("data", subscriptionsForChannel)
            }
        }
    }

    @CrossOrigin
    @GetMapping("/channels/{id}/items")
    fun channelJson(@PathVariable id: String,
                    @RequestParam(required = false) page: Int?,
                    @RequestParam(required = false) pageSize: Int?,
                    @RequestParam(required = false) format: String?,
                    @RequestParam(required = false) q: String?,
                    @RequestParam(required = false) subscriptions: List<String>?,
                    response: HttpServletResponse): ModelAndView {
        return forCurrentUser { user ->
            conditionalLoads.withChannelForUser(id, user) { channel ->
                var mv = ModelAndView(viewFactory.getJsonView())
                if (!Strings.isNullOrEmpty(format) && format == "rss") {    // TODO view factory could do this?
                    mv = ModelAndView(viewFactory.getRssView(channel.name, urlBuilder.getChannelUrl(channel), ""))
                }
                val results = feedItemDAO.getChannelFeedItemsResult(channel, page, q, pageSize, subscriptions)
                feedItemPopulator.populateFeedItems(results, mv, "data")
                val totalCount = results.totalCount
                response.addHeader(X_TOTAL_COUNT, totalCount.toString())
                mv
            }
        }
    }

}